### PR TITLE
Augment analytics so we can answer hardware questions

### PR DESCRIPTION
**Pull request by mikehardy**
## Purpose / Description

- what hardware are we running on?
- what operating system version?
- what screen size?

This came about when I realized the analytics currently did not answer
the question, how many users are using Amazon devices?

While I was in there I added the other hardware bits I thought could
focus development effort allocation

## Approach

Add a method to the initializer for the default analytics request to set device data

Setting User-Agent correctly let's google auto-segment devices to a large degree, I used custom dimensions as well to make sure

## How Has This Been Tested?

API29 emulator, used advanced preferences to turn on debug analytics and watched the analytics console as well as adb logcat to check the request parameters and see their effect on reporting

## Learning (optional, can help others)

The User-Agent style is working code from another project I maintain, so I have confidence in it: 

https://github.com/react-native-community/react-native-device-info/blob/b2d77dbbfc89143747b43f42ae04937d7b2e82ff/android/src/main/java/com/learnium/RNDeviceInfo/RNDeviceModule.java#L725-L735

And the reason to set it, as well as send custom dimensions is because that's just how it's done since there's no API for it:

https://stackoverflow.com/questions/35202625/how-do-i-provide-google-analytics-my-operating-system
https://developers.google.com/analytics/devguides/collection/protocol/v1/parameters

## Checklist

Some unnecessary whitespace changed, specifically auto-format aligned javadoc parameter descriptions, leaving it

- [ ] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/anki/analytics/UsageAnalytics.java [hunk 1]
```diff
@@ -221,9 +228,9 @@ public static void sendAnalyticsEvent(@NonNull String category, @NonNull String
      * Send a detailed arbitrary analytics event, with noun/verb pairs and extra data if needed
      *
      * @param category the category of event, make your own but use a constant so reporting is good
-     * @param action the action the user performed
-     * @param value A value for the event, Integer.MIN_VALUE signifies caller shouldn't send the value
-     * @param label A label for the event, may be null
+     * @param action   the action the user performed
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/analytics/UsageAnalytics.java:231]: These seem erroneous

File: AnkiDroid/src/main/java/com/ichi2/anki/analytics/UsageAnalytics.java [hunk 2]
```diff
@@ -278,4 +285,46 @@ public static void sendAnalyticsException(@NonNull String description, boolean f
...
+    private static class AndroidDefaultRequest extends DefaultRequest {
+        private DefaultRequest setAndroidRequestParameters(Context context) {
+
+            // Are we running on really large screens or small screens or ?
+            try {
+                WindowManager wm = (WindowManager) context.getSystemService(Context.WINDOW_SERVICE);
+                Display display = wm.getDefaultDisplay();
+                Point size = new Point();
+                display.getSize(size);
+                this.screenResolution(size.x + "x" + size.y);
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/analytics/UsageAnalytics.java:304]: Might want to look into how ACRA gets `compatScreenHeightDp`, possibly from `Configuration`

File: AnkiDroid/src/main/java/com/ichi2/anki/analytics/UsageAnalytics.java [hunk 3]
```diff
@@ -278,4 +285,46 @@ public static void sendAnalyticsException(@NonNull String description, boolean f
...
+            } catch (RuntimeException e) {
+                // nothing much to do here, it means we couldn't get WindowManager
+            }
+
+            // We can have up to 20 of these - there might be other things we want to know
+            // but simply seeing what hardware we are running on should be useful
+            this.customDimension(1, Build.VERSION.RELEASE); // systemVersion
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/analytics/UsageAnalytics.java:311]: Got this confused with `BuildConfig.DEBUG`
```suggestion
            this.customDimension(1, Build.VERSION.RELEASE); // readable version: "2.12alpha9"
```

File: AnkiDroid/src/main/java/com/ichi2/anki/analytics/UsageAnalytics.java [hunk 4]
```diff
@@ -278,4 +285,46 @@ public static void sendAnalyticsException(@NonNull String description, boolean f
...
+            this.customDimension(2, Build.BRAND); // brand
+            this.customDimension(3, Build.MODEL); // model
+            this.customDimension(4, Build.BOARD); // deviceId
+
+            // This is important for google to auto-fingerprint us for default reporting
+            try {
+                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
+                    this.userAgent(WebSettings.getDefaultUserAgent(context));
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/analytics/UsageAnalytics.java:319]: Just flagging; I don't think this will trigger #5794

**Comment by david-allison-1** [comment]
I guess it'd be nice to have a few references to the "learning" in the source for this occasion

File: AnkiDroid/src/main/java/com/ichi2/anki/analytics/UsageAnalytics.java (hunk 1 above)
Comment (mikehardy) [reply AnkiDroid/src/main/java/com/ichi2/anki/analytics/UsageAnalytics.java:231]: No, they are fine, they are just textually aligned now fixed-space-horizontal with the beginning of javadoc description for the longest argument 'category', that's what I mentioned before the whitespace-didn't-change thing. Visually it's pleasing. I wouldn't do it manually but the editor did it and it's nice to read when you're reading code so I left it.

File: AnkiDroid/src/main/java/com/ichi2/anki/analytics/UsageAnalytics.java (hunk 3 above)
Comment (mikehardy) [reply AnkiDroid/src/main/java/com/ichi2/anki/analytics/UsageAnalytics.java:311]: No, VERSION.RELEASE is systemVersion, like "10" for Android 10. After maintaining react-native-device-info for 2 years, I know my build constants ;-)

File: AnkiDroid/src/main/java/com/ichi2/anki/analytics/UsageAnalytics.java (hunk 4 above)
Comment (mikehardy) [reply AnkiDroid/src/main/java/com/ichi2/anki/analytics/UsageAnalytics.java:319]: I/we (at react-native-device-info) catch RuntimeException on purpose so it never crashes. Nearly 10 million downloads on that piece of code with no reported crashes, I'm comfortable with this block

**Comment by mikehardy** [comment]
> There's a lot of whitespace changes which appear to be erroneous.

as explained, just text horizontal align vs longest argument name in javadoc. harmless

> Where are the current dimensions? I'd have thought Android Version to be a good dimension if it's not a default.

System version is android version which we will get now in a custom dimension, plus now that I'm sending the User-Agent as generated by the WebView analytics will likely auto-parse it for use in their standard reporting (it is in there as well but user-agents are unreliable and they could change it, so custom dimension also)

There were no custom dimensions prior, here was the default request stuff prior:

https://github.com/mikehardy/google-analytics-java/blob/d699c8e68f0f4710f1e2e4657c2d8308a2987b06/src/main/java/com/brsanthu/googleanalytics/discovery/DefaultRequestParameterDiscoverer.java#L31-L46

Just region, language, document-encoding, and a User-Agent that was effectively useless since Google has custom user-agent parsing libraries for auto-segmentation and this one didn't fit anything for them and thus was effectively discarded

**Comment by mikehardy** [comment]
extended comments

File: AnkiDroid/src/main/java/com/ichi2/anki/analytics/UsageAnalytics.java (hunk 3 above)
Comment (david-allison-1) [reply AnkiDroid/src/main/java/com/ichi2/anki/analytics/UsageAnalytics.java]: 🤦 Definitely needed those comments then, thanks!

**Comment by mikehardy** [comment]
This is related both to the Amazon App Store work (I'm genuinely curious about Kindle Fire users) plus tangentially to the Xiaomi-cloze-failure case (Vietnam appears to be at least 50% Xiaomi and they're vociferous about cloze not working, how many are there really?)

**Pull request merged**